File: 677_map_sum/map_sum.go
/*
 * @Date: 2021-11-14 02:00:44
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-11-14 02:10:28
 */

package main

import "strings"

type MapSum map[string]int

func Constructor() MapSum {
	return MapSum{}
}

func (m MapSum) Insert(key string, val int) {
	m[key] = val
}

func (m MapSum) Sum(prefix string) (sum int) {
	for s, v := range m {
		if strings.HasPrefix(s, prefix) {
			sum += v
		}
	}
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("assertion failed")
		}
	}
	ms := Constructor()
	ms.Insert("apple", 3)
	assert(ms.Sum("ap") == 3)
	ms.Insert("app", 2)
	assert(ms.Sum("ap") == 5)
}

File: cpp/677_map_sum/map_sum.cpp
/*
 * @Date: 2021-11-14 02:00:40
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-11-14 02:09:02
 */

#include <cassert>
#include <string>
#include <unordered_map>

using namespace std;

class MapSum {
 public:
  MapSum() {}

  void insert(string key, int val) { cnt[key] = val; }

  int sum(string prefix) {
    int res = 0;
    for (auto& [key, val] : cnt) {
      if (key.substr(0, prefix.size()) == prefix) {
        res += val;
      }
    }
    return res;
  }

 private:
  unordered_map<string, int> cnt;
};

int main() {
  MapSum ms = MapSum();
  ms.insert("apple", 3);
  assert(ms.sum("ap") == 3);
  ms.insert("app", 2);
  assert(ms.sum("ap") == 5);
}